File: corpus/deploy/docker/Dockerfile.scala
# =============================================================================
# Obraz analytics-pipeline (Scala 2.13, Spark 3.5).
#
# Staví se tlustý jar přes sbt-assembly, protože Spark potřebuje jeden
# artefakt, který rozešle executorům. Vyloučené jsou Spark samotný a Hadoop —
# ty jsou v základním obrazu a duplicitní verze v jaru způsobovala
# NoSuchMethodError až za běhu na executoru, ne při startu driveru.
#
# Obraz slouží dvěma věcem naráz: jako driver (Deployment, obsluhuje
# GET /v1/metrics/*) a jako executor, kterého si driver spouští přes
# Kubernetes API. Proto tu zůstává celý Spark runtime, i když ho HTTP část
# nepotřebuje — dva obrazy by znamenaly dvě verze, které se můžou rozejít.
# =============================================================================

FROM sbtscala/scala-sbt:eclipse-temurin-jammy-21.0.2_13_1.9.9_2.13.13 AS build

WORKDIR /src
COPY services/analytics/build.sbt ./
COPY services/analytics/project ./project
RUN --mount=type=cache,target=/root/.cache/coursier \
    --mount=type=cache,target=/root/.sbt \
    sbt update

COPY libs/ /src/libs/
COPY services/analytics/src ./src
RUN --mount=type=cache,target=/root/.cache/coursier \
    --mount=type=cache,target=/root/.sbt \
    sbt assembly

FROM apache/spark:3.5.1-scala2.12-java17-python3-ubuntu AS runtime

ARG VERSION=4.2.0
ARG BUILD_SHA=unknown
LABEL org.opencontainers.image.title="analytics-pipeline" \
      org.opencontainers.image.version="${VERSION}" \
      org.opencontainers.image.revision="${BUILD_SHA}" \
      org.opencontainers.image.vendor="ORBITALFREIGHT"

USER root
RUN groupadd -g 1400 orbitalfreight \
    && useradd -u 1400 -g 1400 -s /usr/sbin/nologin -M orbitalfreight \
    && mkdir -p /var/tmp/spark /var/lib/orbitalfreight/routing \
    && chown -R 1400:1400 /var/tmp/spark /var/lib/orbitalfreight/routing

COPY --from=build --chown=1400:1400 /src/target/scala-2.13/analytics-pipeline-assembly.jar /opt/orbitalfreight/analytics.jar
COPY --chown=1400:1400 deploy/docker/conf/analytics-job /opt/orbitalfreight/bin/analytics-job

USER 1400:1400
EXPOSE 8093

# Driver čte přes roli of_analytics_ro (OF_ANALYTICS_READONLY_DATABASE_URL) ze
# všech schémat a zapisuje výhradně do `analytics` druhým připojením. Sloučit
# je do jednoho by dávkové úloze dalo právo zapisovat do cizích schémat.
ENTRYPOINT ["/opt/entrypoint.sh"]
CMD ["driver", "--class", "net.orbitalfreight.analytics.Main", "/opt/orbitalfreight/analytics.jar"]

File: corpus/deploy/docker/Dockerfile.go
# =============================================================================
# Obraz pro obě služby psané v Go 1.22: identity-service a audit-ledger.
#
# Jeden soubor pro obě schválně — liší se jen adresářem a portem, a dvě kopie
# téhož by znamenaly dvě místa, kde se zapomene na opravu. Která služba se
# staví, říká ARG SERVICE; hodnota musí být jméno ze §1 SPEC.md, protože se
# z ní odvozuje i cesta do services/.
#
# Výsledný obraz je distroless a běží pod UID 1400 — stejným, jaké má
# securityContext v deploy/kubernetes/. Kdyby se rozešly, pody by padaly na
# právech k připojeným secretům, ne na aplikační chybě.
#
#   docker build -f deploy/docker/Dockerfile.go --build-arg SERVICE=identity-service .
# =============================================================================

FROM golang:1.22-bookworm AS build

ARG SERVICE
ARG VERSION=4.2.0
ARG BUILD_SHA=unknown

WORKDIR /src

# Závislosti odděleně od zdrojáků: mění se řádově méně často a vrstva se tak
# recykluje mezi buildy všech Go služeb.
COPY go.mod go.sum ./
RUN --mount=type=cache,target=/go/pkg/mod \
    go mod download

COPY libs/ ./libs/
COPY services/${SERVICE}/ ./services/${SERVICE}/

# CGO vypnuté kvůli distroless static; -trimpath a prázdné ldflags kvůli
# reprodukovatelnosti — dva buildy téhož commitu musí dát stejný digest.
# Verze a SHA se vpisují do binárky a vrací je GET /version (§3.15).
RUN --mount=type=cache,target=/go/pkg/mod \
    --mount=type=cache,target=/root/.cache/go-build \
    CGO_ENABLED=0 GOOS=linux go build \
      -trimpath \
      -ldflags="-s -w -X main.version=${VERSION} -X main.buildSHA=${BUILD_SHA}" \
      -o /out/service \
      ./services/${SERVICE}/cmd/server

FROM gcr.io/distroless/static-debian12:nonroot AS runtime

ARG SERVICE
ARG VERSION=4.2.0
ARG BUILD_SHA=unknown

LABEL org.opencontainers.image.title="${SERVICE}" \
      org.opencontainers.image.version="${VERSION}" \
      org.opencontainers.image.revision="${BUILD_SHA}" \
      org.opencontainers.image.vendor="ORBITALFREIGHT" \
      org.opencontainers.image.source="https://git.orbitalfreight.net/platform/orbitalfreight"

COPY --from=build /out/service /usr/local/bin/service

# 1400 = uživatel orbitalfreight, stejný jako v rolích Ansiblu i v podech.
USER 1400:1400

# identity-service 8081/9081, audit-ledger 8092/9092. EXPOSE je dokumentace,
# skutečný port říká OF_HTTP_PORT.
EXPOSE 8081 9081 8092 9092

ENTRYPOINT ["/usr/local/bin/service"]
